### PR TITLE
[SPARK-31079][SQL] Logging QueryExecutionMetering in RuleExecutor logger

### PR DESCRIPTION
**Pull request by Eric5553**
### What changes were proposed in this pull request?
RuleExecutor already support metering for analyzer/optimizer rules. By providing such information in `PlanChangeLogger`, user can get more information when debugging rule changes .

This PR enhanced `PlanChangeLogger` to display RuleExecutor metrics. This can be easily done by calling the existing API `resetMetrics` and `dumpTimeSpent`, but there might be conflicts if user is also collecting total metrics of a sql job. Thus I introduced `QueryExecutionMetrics`, as the snapshot of `QueryExecutionMetering`, to better support this feature.

Information added to `PlanChangeLogger`
```
=== Metrics of Executed Rules ===
Total number of runs: 554
Total time: 0.107756568 seconds
Total number of effective runs: 11
Total time of effective runs: 0.047615486 seconds
```

### Why are the changes needed?
Provide better plan change debugging user experience

### Does this PR introduce any user-facing change?
Only add more debugging info of `planChangeLog`, default log level is TRACE.

### How was this patch tested?
Update existing tests to verify the new logs

### CODE REVIEW

**Comment by Eric5553** [comment]
retest this please

**Comment by maropu** [comment]
Are these metrics benefit for general users? If these metrics are for developers, how about just showing these metrics via `spark.sql.optimizer.planChangeLog.level`?

**Comment by Eric5553** [comment]
> Are these metrics benefit for general users? If these metrics are for developers, how about just showing these metrics via a logger with `spark.sql.optimizer.planChangeLog.level`?

Yea, logging message should be enough. I'll make the change, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 1]
```diff
@@ -231,7 +238,7 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
             s"""
                |=== Result of Batch ${batchName} ===
                |${sideBySide(oldPlan.treeString, newPlan.treeString).mkString("\n")}
-          """.stripMargin
+            """.stripMargin
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: unnecessary change?

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: I thought the padding was incorrect, so fixed it here together.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: ok

**Comment by maropu** [comment]
The improvement looks fine to me.  WDYT? @maryannxue @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerLoggingSuite.scala [hunk 2]
```diff
@@ -49,12 +49,19 @@ class OptimizerLoggingSuite extends PlanTest {
       case event => Seq(
         "Applying Rule",
         "Result of Batch",
-        "has no effect").exists(event.getRenderedMessage().contains)
+        "has no effect",
+        "Metrics of Executed Rules").exists(event.getRenderedMessage().contains)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerLoggingSuite.scala]: this only exists if `Result of Batch` exists?

Comment (Eric5553) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerLoggingSuite.scala]: The metrics will be reported at `RuleExecutor.execute` level. I double checked the code and found I logged it at the wrong place within `batches.foreach` loop.  Fixed it by abc864ef7e41d65932060436331a57db8d77fbe7.

**Comment by Eric5553** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala [hunk 3]
```diff
@@ -95,3 +107,26 @@ case class QueryExecutionMetering() {
...
   }
 }
+
+case class QueryExecutionMetrics(
+    time: Long,
+    numRuns: Long,
+    numEffectiveRuns: Long,
+    timeEffective: Long) {
+
+  def +(metrics: QueryExecutionMetrics): QueryExecutionMetrics = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala]: is this method needed?

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala]: Emm. Not used by now, agree to add it once we really need this. Removed in 36bf2e134436a9b5c7d96c839855dd69cef45538.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by Eric5553** [comment]
@cloud-fan @maropu Thanks so much!

**Comment by maropu** [comment]
late LGTM
